### DISCUSSION TITLE
WeightedSwitchController vs. RandomSwitchController

### DISCUSSION THREAD

**Discussion by MaksSieve** [abstracta/jmeter-java-dsl#96] [Ideas] [answered]
Hi!

While working on integration BlazeMeter's WeightedSwitchController to dsl, I've also realised that there is no analogs for Gatling's `randomSwitch` ([link](https://github.com/gatling/gatling/blob/main/gatling-core/src/main/scala/io/gatling/core/structure/ConditionalStatements.scala#L120)). 

And I've implement it too :)

But then I've understood that in fact they have basically same logic: 

* WheightedSwitchConroller selects child by checking total sum of all iterations ([realisation](https://github.com/MaksSieve/jmeter-java-dsl/blob/weighted_switch_controller/jmeter-java-dsl/src/main/java/us/abstracta/jmeter/javadsl/core/controllers/DslWeightedSwitchController.java))
* my realisation of randomSwitch it tosses a coin (ThreadLocalRandom) every turn ([realisation](https://github.com/MaksSieve/jmeter-java-dsl/blob/random_switch_controller/jmeter-java-dsl/src/main/java/us/abstracta/jmeter/javadsl/core/controllers/DslRandomSwitchController.java))

Finally, they are really close to be the same.

So, I'd like to you estimate both and choose what is better for dsl, may be they deserve to exist together :)

#93

**Comment by rabelenda** [comment, marked as answer]
Hello, 

I don't see in your implementation of randomSwitch the actual import for RandomSwitchController class. Is this jmeter RandomController, or is it WeightedSwitchController with "Random Choice" option selected, or is it some other controller implementation?

I think, that if you implement WeightedSwitchController, you might want to consider adding to it a `randomSelection` method to allow for setting the "Random Choice" option.

Now, considering that, jmeter RandomController would be similar to weightedSwitchController with random choice and all elements with same weight (or no weight). Ie:

```java
//in this scenario we are not using add(double, ThreadGroupChild) since there is no need to set weights in children.
weightedSwitchController()
  .randomSelection(true)
  .children(
     httpSampler("https://myservice"),
     httpSampler("https://myservice/users")
  )
```

But, I think that for brevity, and avoiding some confusion that may rise with previous case, would be nice to have a shortcut, that is easily identifiable by jmeter users like:

```java
randomSwitchController(
   httpSampler("https://myservice"),
   httpSampler("https://myservice/users")
)
```

Does this answer your question? What are your thoughts?

In any case, I would advise you to do a PR for one, and then another for the other one if you plan to implement both.

**Reply by MaksSieve** [reply]
Thx, for the fast response!

my RandomSwitch controller implements this logic:
1. on configuration it saves children, their probabilities and generate **_bounds_** - special array that contains the right bounds of each child's probability range.
2. when it is called it toss a coin (generate number from 1 to 100)
3. then it seek the child whose range the coin fell in, in case the coin fell in empty range (always the left one - from 0 to the first real child bound) it choose to skip this iteration
4. then it pass that children to jmeter

so it's the brand new conrtoller that doesn't exist in nature before)

but after i implemented it i found that in fact it copies the behaviour of blazemeter's WeightedSwitchController, because i.e. if we have conroller with this settings:
```java
randomSwitchController()
  .add(10, httpSampler("https://myservice"))
  .add(20, httpSampler("https://myservice/users"))
``` 
it equals this configuration of weighted controller:
```java
weightedSwitchController()
  .add(1, httpSampler("https://myservice"))
  .add(2, httpSampler("https://myservice/users"))
```

so I have the another implementation with the same logic but in my opinion much understandable :)

**Reply by rabelenda** [reply]
your implementation sounds like a mix of percentController & weightedSwitchController, because some times skips iterations.

I think your first example would be translated to this:

```java
percentController(30, 
  weightedSwitchController()
     .add(1, httpSampler("https://myservice"))
     .add(2, httpSampler("https://myservice/users"))
) 
```

The main difference from your implementation and above, if I understood your implementation behavior description correctly, would be that percentController is deterministic (you can check PercentControler javadoc for an explanation of what I mean) while your implementation would be non deterministic.

If is only that the difference, I would consider using percentController + weightedSwitchController to achieve the behavior you want, instead of adding a new concept to the DSL. 

Eventually, if we consider it is worth to have a non deterministic percentController (which, with the info I currently have, I don't think is worth, but may change in the future if new info is provided), we might add a method in percentController to allow for non deterministic resolution and with such method change the controller implementation, keeping DSL API simple (no additional controller to worry for end users) but allowing to switch it's internal behavior (which in turns uses a different controller in JMeter).

Regards and thank you for this interesting exchange.

**Reply by MaksSieve** [reply]
Thanks for the detailed response. You are right, my implementations is a sort of indeterministic PercentController. The key difference is that if I have sequential PercentControllers like this:
```java
...
percentController(50, httpSampler("https://myservice"))
percentController(25, httpSampler("https://myservice/user"))
percentController(12.5, httpSampler("https://myservice/serve"))
...
```
each of them will be executed in sequence and in some case there will appear all three requests:
```
1 - [0,1,0,1,0,1,0,1]
2 - [0,0,0,1,0,0,0,1]
3 - [0,0,0,0,0,0,0,1]
```
But I wish to have controller, that on every time it fires, chooses only one with specific probability. 

And this logic is the same that WeightedController has, the difference is only internal implementation)
So, I will finish work on wheightedSwitchController and send a PR in short time!

Thanks for the discussion:)
